include/vara/Feature/Feature.h

```diff
@@ -203,34 +136,90 @@ class Feature {
+  bool operator!=(const vara::feature::Feature &F) const {
+    return !operator==(F);
+  }
+
+  /// Compare in depth first ordering.
+  bool operator<(const vara::feature::Feature &F) const;
+  bool operator>(const vara::feature::Feature &F) const {
+    return F.operator<(*this);
+  }
+
+  [[nodiscard]] FeatureSourceRange *getFeatureSourceRange() {
     return Loc.has_value() ? &Loc.value() : nullptr;
   }
 
   //===--------------------------------------------------------------------===//
   // Utility
   [[nodiscard]] virtual std::string toString() const;
 
-  bool operator==(const Feature &F) const {
-    return llvm::StringRef(this->getName()).lower() ==
-           llvm::StringRef(F.getName()).lower();
-  }
+  void print(std::ostream &Out) const { Out << toString() << std::endl; }
+  void print(llvm::raw_ostream &Out) const { Out << toString() << '\n'; }
 
-  bool operator<(const Feature &F) const {
-    return llvm::StringRef(this->getName()).lower() <
-           llvm::StringRef(F.getName()).lower();
-  }
+  LLVM_DUMP_METHOD
+  void dump() const { llvm::outs() << toString() << "\n"; }
+
+protected:
+  Feature(FeatureType T, string Name, bool Opt,
+          std::optional<FeatureSourceRange> Loc, Feature *Parent,
+          FeatureSetType Children, FeatureSetType Excludes,
+          FeatureSetType Implications, FeatureSetType Alternatives)
+      : T(T), Name(std::move(Name)), Opt(Opt), Loc(std::move(Loc)),
+        Parent(Parent), Children(std::move(Children)),
+        Excludes(std::move(Excludes)), Implications(std::move(Implications)),
+        Alternatives(std::move(Alternatives)) {}
+
+private:
+  friend class FeatureModel;
+  friend class FeatureModelBuilder;
+
+  FeatureType T;
+  string Name;
+  bool Opt;
+  std::optional<FeatureSourceRange> Loc;
+  Feature *Parent;
+  FeatureSetType Children;
+  FeatureSetType Excludes;
+  FeatureSetType Implications;
+  FeatureSetType Alternatives;
+
+  void addChild(Feature *F) { Children.insert(F); }
+
+  void setParent(Feature *F) { Parent = F; }
+
+  void addExclude(Feature *F) { Excludes.insert(F); }
+
+  void addAlternative(Feature *F) { Alternatives.insert(F); }
+
+  void addImplication(Feature *F) { Implications.insert(F); }
 };
 
 /// Options without arguments.
 class BinaryFeature : public Feature {
 
 public:
-  BinaryFeature(string Name, bool Opt,
-                std::optional<Location> Loc = std::nullopt)
-      : Feature(std::move(Name), Opt, std::move(Loc)) {}
+  BinaryFeature(string Name, bool Opt = false,
+                std::optional<FeatureSourceRange> Loc = std::nullopt,
+                Feature *Parent = nullptr, FeatureSetType Children = {},
+                FeatureSetType Excludes = {}, FeatureSetType Implications = {},
+                FeatureSetType Alternatives = {})
+      : Feature(BINARY, std::move(Name), Opt, std::move(Loc), Parent,
+                std::move(Children), std::move(Excludes),
+                std::move(Implications), std::move(Alternatives)) {}
```

Comment:
FM consistency is assured through that Features will only be added with `FeatureModel::addFeature(std::unique_ptr<Feature>) : bool` checking whether a Feature can be inserted without violating any rules (currently only rudimentary). On the other hand it may be handy to create (unrelated) features beforehand to store information, building the actual FM afterwards (as the builder provides corresponding methods).